fiona/_geometry.pyx

```diff
@@ -101,66 +110,150 @@ cdef class GeomBuilder:
         parts = []
-        for j in range(OGR_G_GetGeometryCount(geom)):
+        j = 0
+        count = OGR_G_GetGeometryCount(geom)
+        while j < count:
             part = OGR_G_GetGeometryRef(geom, j)
-            parts.append(GeomBuilder().build(part))
+            code = base_geometry_type_code(OGR_G_GetGeometryType(part))
+            if code in (
+                OGRGeometryType.PolyhedralSurface.value,
+                OGRGeometryType.TIN.value,
+                OGRGeometryType.Triangle.value,
+            ):
+                OGR_G_RemoveGeometry(geom, j, False)
+                # Removing a geometry will cause the geometry count to drop by one,
+                # and all “higher” geometries will shuffle down one in index.
+                count -= 1
+                parts.append(GeomBuilder().build(part))
+            else:
+                parts.append(GeomBuilder().build(part))
+                j += 1
         return parts
 
-    cpdef _buildPolygon(self):
-        coordinates = [p['coordinates'] for p in self._buildParts(self.geom)]
+    cdef dict _buildPolygon(self, void *geom):
+        coordinates = [p['coordinates'] for p in self._buildParts(geom)]
         return {'type': 'Polygon', 'coordinates': coordinates}
 
-    cpdef _buildMultiPoint(self):
-        coordinates = [p['coordinates'] for p in self._buildParts(self.geom)]
+    cdef dict _buildMultiPoint(self, void *geom):
+        coordinates = [p['coordinates'] for p in self._buildParts(geom)]
         return {'type': 'MultiPoint', 'coordinates': coordinates}
 
-    cpdef _buildMultiLineString(self):
-        coordinates = [p['coordinates'] for p in self._buildParts(self.geom)]
+    cdef dict _buildMultiLineString(self, void *geom):
+        coordinates = [p['coordinates'] for p in self._buildParts(geom)]
         return {'type': 'MultiLineString', 'coordinates': coordinates}
 
-    cpdef _buildMultiPolygon(self):
-        coordinates = [p['coordinates'] for p in self._buildParts(self.geom)]
+    cdef dict _buildMultiPolygon(self, void *geom):
+        coordinates = [p['coordinates'] for p in self._buildParts(geom)]
         return {'type': 'MultiPolygon', 'coordinates': coordinates}
 
-    cpdef _buildGeometryCollection(self):
-        parts = self._buildParts(self.geom)
+    cdef dict _buildGeometryCollection(self, void *geom):
+        parts = self._buildParts(geom)
         return {'type': 'GeometryCollection', 'geometries': parts}
 
-    cdef build(self, void *geom):
-        # The only method anyone needs to call
-        if geom == NULL:
-            raise ValueError("Null geom")
 
-        cdef unsigned int etype = OGR_G_GetGeometryType(geom)
+    cdef object build_from_feature(self, void *feature):
+        # Build Geometry from *OGRFeatureH
+        cdef void *cogr_geometry = NULL
+        cdef int code
+
+        cogr_geometry = OGR_F_GetGeometryRef(feature)
+        code = base_geometry_type_code(OGR_G_GetGeometryType(cogr_geometry))
```

Comment:
Currently, the code of a geometry is determined two times: here and in build(). This could be optimized by passing the code to build.  What is the equivalent in Cython of build(..., code: Optional[int] = None)? Passing an int pointer that is NULL when no code was previously computed?

Reply:
@rbuffat don't worry about optimizing that. Getting the code is practically free compared to other stuff we're doing.